app/models/miq_ae_domain.rb

```diff
@@ -8,25 +10,55 @@ class MiqAeDomain < MiqAeNamespace
   EDITABLE_PROPERTIES_FOR_REMOTES = [:priority, :enabled].freeze
   AUTH_KEYS = %w[userid password].freeze
 
-  default_scope { roots.where(arel_table[:name].not_eq("$")) }
+  default_scope { roots.where.not(:name => "$") }
+  scope :all_domains, -> { unscoped.roots }
   validates :ancestry, :inclusion => {:in => [nil], :message => 'should be nil for Domain'}
 
   validates_presence_of :tenant, :message => "object is needed to own the domain"
   after_destroy :squeeze_priorities
+  after_destroy :clear_domain_cache
+  after_save :clear_domain_cache
   default_value_for :source,  USER_SOURCE
   default_value_for :enabled, false
   before_save :default_priority
   belongs_to :tenant
   belongs_to :git_repository, :dependent => :destroy
   validates_inclusion_of :source, :in => VALID_SOURCES
 
+  alias_attribute :domain_name, :name
+
   EXPORT_EXCLUDE_KEYS = [/^id$/, /^(?!tenant).*_id$/, /^created_on/, /^updated_on/,
                          /^updated_by/, /^reserved$/, /^commit_message/,
                          /^commit_time/, /^commit_sha/, /^ref$/, /^ref_type$/,
                          /^last_import_on/].freeze
 
   include TenancyMixin
 
+  def self.name_to_id(name)
+    return if name.nil?
+
+    load_domain_cache
+    name = name.downcase
+    @domain_names[name] ||= all_domains.where(:lower_name => name).pluck(:id).first
+  end
+
+  def self.id_to_name(domain_id)
+    return if domain_id.nil?
+
+    load_domain_cache
+    @domain_ids[domain_id] ||= all_domains.where(:id => domain_id).pluck(:name).first
+  end
+
+  delegate :clear_domain_cache, :to => self
+  def self.clear_domain_cache
+    @domain_ids = @domain_names = nil
+  end
+
+  def self.load_domain_cache
+    @domain_ids ||= Hash[*all_domains.pluck(:id, :name).flatten]
+    @domain_names ||= @domain_ids.invert.transform_keys { |key| key.to_s.downcase }
+  end
```

Comment:
Are these caches shared across threads?  Since this is being manipulated as part of a callback, I'm concerned that there might be some race condition with the clearing and loading of caches.

Reply:
Additinoally, what happens when someone, say, changes a domain in the UI (which will clear the cache in the UI worker)...how does that cache busting propogate to the actual automate workers where it matters?

Reply:
Automate is not thread safe. So across threads is currently not a problem, though assumptions like this in hidden components tend to rear their heads at inopportune times.

Also of note, A domain name will probably never change name and will rarely be deleted or added.
If the domain name is deleted, that is not a problem, the id will not be looked up to convert into a name.
If one is added, this will look it up.
If one is changed, I think the full text name/url will be stored in the cache and passed around. This is not a good situation regardless of caching.